src/tools/rbd_mirror/ImageReplayer.cc

```diff
@@ -961,25 +961,27 @@ template <typename I>
 void ImageReplayer<I>::handle_replayer_notification() {
   dout(10) << dendl;
 
-  // detect a rename of the local image
-  std::string local_image_name;
-  {
-    std::shared_lock image_locker{m_local_image_ctx->image_lock};
-    local_image_name = m_local_image_ctx->name;
+  std::unique_lock locker{m_lock};
+  if (m_state != STATE_REPLAYING) {
+    // might be attempting to shut down
+    return;
   }
 
   {
-    std::unique_lock locker{m_lock};
-    ceph_assert(m_state == STATE_REPLAYING);
-    ceph_assert(m_replayer != nullptr);
-
-    if (m_local_image_name != local_image_name) {
+    // detect a rename of the local image
+    ceph_assert(m_local_image_ctx != nullptr);
+    std::shared_lock image_locker{m_local_image_ctx->image_lock};
+    if (m_local_image_name != m_local_image_ctx->name) {
       // will re-register with new name after next status update
       dout(10) << "image renamed" << dendl;
-      m_local_image_name = local_image_name;
+      m_local_image_name = m_local_image_ctx->name;
     }
   }
 
+  // replayer cannot shut down until after callback completes
```

Comment:
@dillaman Was it intentional to add this comment and leave the one below? They seem to me very similar?

Reply:
Missed the second one